airbyte-cdk/python/unit_tests/sources/declarative/interpolation/test_jinja.py

```diff
@@ -138,7 +137,39 @@ def test_negative_day_delta():
     ],
 )
 def test_macros(s, expected_value):
-    interpolation = JinjaInterpolation()
     config = {}
     val = interpolation.eval(s, config)
     assert val == expected_value
+
+
+@pytest.mark.parametrize(
+    "template_string", [
+        pytest.param("{{ import os) }}", id="test_jinja_with_import"),
+        pytest.param("{{ [a for a in range(1000000000)] }}", id="test_jinja_with_list_comprehension"),
+    ]
+)
+def test_invalid_jinja_statements(template_string):
+    config = {"key": "value"}
+    with pytest.raises(TemplateSyntaxError):
+        interpolation.eval(template_string, config=config)
+
+
+@pytest.mark.parametrize("template_string", [
+    # This test stalls if range is removed from JinjaInterpolation.RESTRICTED_BUILTIN_FUNCTIONS
+    pytest.param("""
+       {% set a = 1 %}
+       {% set b = 1 %}
+       {% for i in range(1000000000) %}
+       {% endfor %}
+        {{ a }}""", id="test_jinja_with_very_long_running_compute"),
+    pytest.param("{{ eval ('2+2') }}", id="test_jinja_with_eval"),
+    pytest.param("{{ getattr(config, 'key') }}", id="test_getattr"),
+    pytest.param("{{ setattr(config, 'password', 'hunter2') }}", id="test_setattr"),
+    pytest.param("{{ globals()  }}", id="test_jinja_with_globals"),
+    pytest.param("{{ locals()  }}", id="test_jinja_with_globals"),
+    pytest.param("{{ eval ('2+2') }}", id="test_jinja_with_eval"),
+])
+def test_restricted_builtin_functions_are_not_executed(template_string):
```

Comment:
is there a way to make this raise exceptions?

Reply:
No because they're not special keywords to Jinja. I just added the tests to confirm that assumption

Reply:
as per in person chat, we might be able to add a condition during evaluation for disallowed phrases like `in range(` and throw a ValueError if the interpolation string contains this since its invalid. We'll still remove it from the context for extra safety, but this could fail more loudly

Reply:
Updated [the `JinjaInterpolation` to raise an error if a variable is undefined](https://github.com/airbytehq/airbyte/pull/25589/files#diff-df8a64d2d501197620f175b1b69fcbb3ad0117f621cfe3fbc387b4e43b39f34aR94-R98) instead of blindly returning `None`